docs/src/lib/indexing.md

```diff
@@ -138,15 +138,26 @@ so it is unsafe to use it afterwards (the column length correctness will be pres
                        `v` must be an `AbstractMatrix` or an `AbstractDataFrame`
                        (in the latter case column names must match);
 
-`setindex!` on `SubDataFrame`:
+`setindex!` on `SubDataFrame` (not created with `:` as column selector):
 * `sdf[row, col] = v` -> set value of `col` in row `row` to `v` in-place;
 * `sdf[CartesianIndex(row, col)] = v` -> the same as `sdf[row, col] = v`;
 * `sdf[row, cols] = v` -> the same as `dfr = df[row, cols]; dfr[:] = v` in-place;
 * `sdf[rows, col] = v` -> set rows `rows` of column `col`, in-place; `v` must be an abstract vector;
 * `sdf[rows, cols] = v` -> set rows `rows` of columns `cols` in-place;
                            `v` can be an `AbstractMatrix` or `v` can be `AbstractDataFrame` when column names must match;
-
-Note that `sdf[!, col] = v`, `sdf[!, cols] = v` and `sdf.col = v` are not allowed as `sdf` can be only modified in-place.
+* `sdf[!, col] = v` -> replaces `col` with `v` with copying; if `col` is present if `sdf`
+                       then filtered-out rows in newly created vector are filled with
+                       values already present in that column;
+                       if `col` is not present in `sdf` then the operation is only allowed
+                       if `sdf` was created with `:` as column selector. In this case
+                       filtered-out rows are filled with `missing`;
+                      equivalent to `sdf.col = v` if `col` is a valid identifier;
+                      operation is allowed if `length(v) == nrow(sdf)`;
+* `sdf[!, cols] = v` -> replaces existing columns `cols` in data frame `sdf` with copying;
```

Comment:
Do you mean `sdf[:, cols]` rather than `sdf[!, cols]`?

I'm not sure what's being copied here. is `v` needs to get new memory allocated regardless, right? Is it that the entire column `df.v` is replaced? That can't be fast, if someone is doing a loop with multiple groups.

Reply:
I think `!` is what is meant here. But indeed docs for `:` are missing AFAICT.

Reply:
> But indeed docs for `:` are missing AFAICT.

In this case `:` is just `rows` and it is covered in `sdf[rows, cols] = v` rule.

> I'm not sure what's being copied here. (...) That can't be fast, if someone is doing a loop with multiple groups.

This is the point of difference of `:` and `!`:
* `!` makes a copy and makes a promotion.
* `:` is in-place and makes a conversion.

See this example that hopefully makes it clear:
```
julia> df = DataFrame(ones(3,2), :auto)
3×2 DataFrame
 Row │ x1       x2      
     │ Float64  Float64 
─────┼──────────────────
   1 │     1.0      1.0
   2 │     1.0      1.0
   3 │     1.0      1.0

julia> dfv = @view df[2:2, :]
1×2 SubDataFrame
 Row │ x1       x2      
     │ Float64  Float64 
─────┼──────────────────
   1 │     1.0      1.0

julia> dfv[:, :] = ['a' 'a']
1×2 Matrix{Char}:
 'a'  'a'

julia> df
3×2 DataFrame
 Row │ x1       x2      
     │ Float64  Float64 
─────┼──────────────────
   1 │     1.0      1.0
   2 │    97.0     97.0
   3 │     1.0      1.0

julia> df = DataFrame(ones(3,2), :auto)
3×2 DataFrame
 Row │ x1       x2      
     │ Float64  Float64 
─────┼──────────────────
   1 │     1.0      1.0
   2 │     1.0      1.0
   3 │     1.0      1.0

julia> dfv = @view df[2:2, :]
1×2 SubDataFrame
 Row │ x1       x2      
     │ Float64  Float64 
─────┼──────────────────
   1 │     1.0      1.0

julia> dfv[!, :] = ['a' 'a']
1×2 Matrix{Char}:
 'a'  'a'

julia> df
3×2 DataFrame
 Row │ x1   x2  
     │ Any  Any 
─────┼──────────
   1 │ 1.0  1.0
   2 │ a    a
   3 │ 1.0  1.0
```

Reply:
Ah I see that makes sense. 

This may end up being a real pitfall, performance wise. Someone might do 

```
for ai in unique(df.a)
    sdf = subset(df, :a => ==(ai))
    sdf[!, :b] = 100
end
```

If `unique(df.a)` is large and `nrow(df)` is large, this will cause horrendous allocations, right?

Reply:
Yes - it will allocate a lot. If one is sure using in-place operation is safe then `sdf[:, :b]` should be used.
The timing will not be super bad, but it will be noticeably worse than `sdf[:, :b]`.